federation/federation.go

```diff
@@ -0,0 +1,195 @@
+package federation
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var (
+	headSnapshotKey = datastore.NewKey("head-snapshot")
+	logger          = logging.Logger("federation")
+)
+
+type (
+	Federation struct {
+		*options
+		members []peer.ID
+		vc      vectorClock
+
+		mu                   sync.Mutex
+		snapshotTicker       *time.Ticker
+		reconciliationTicker *time.Ticker
+		headNodeCache        ipld.Node
+		ctx                  context.Context
+		cancel               context.CancelFunc
+	}
+)
+
+func New(o ...Option) (*Federation, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Federation{
+		options:              opts,
+		snapshotTicker:       time.NewTicker(opts.snapshotInterval),
+		reconciliationTicker: time.NewTicker(opts.reconciliationInterval),
+		vc:                   newVectorClock(),
+		ctx:                  ctx,
+		cancel:               cancel,
+	}, nil
+}
+
+func (f *Federation) Start(_ context.Context) error {
```

Comment:
Why take and not use cancel here?

Reply:
Because it might be surprising to the user that start context is tied to the entire lifecycle of federation rather than just starting it.